pageserver/src/tenant.rs

```diff
@@ -447,6 +449,8 @@ pub enum DeleteTimelineError {
     NotFound,
     #[error("HasChildren")]
     HasChildren,
+    #[error("stop upload queue: {0:#}")]
```

Comment:
What does this error mean? Is it expected in some scenarios?

Reply:
Yeah, see the place where we inspect the `.stop()` return value:

```rust
// Prevent new uploads from starting.
        if let Some(remote_client) = timeline.remote_client.as_ref() {
            let res = remote_client.stop();
            match &res {
                Ok(()) => {}
                Err(e) => match e {
                    remote_timeline_client::StopError::QueueUninitialized => {
                        // This could happen if the timeline is Broken, e.g., because it failed to fetch IndexPart when it was loaded.
                    }
                },
            }
            res?;
        }
    ```


I'll remove the `#[from]` so that a simple code search for `StopUploadQueue` will lead to the place where we bubble it up.

Reply:
https://github.com/neondatabase/neon/pull/3919/commits/464bd27053c81f182db844e642ce086a3908ba80

Reply:
It's still not 100% clear to me. I understand these error codes something like this:

NotFound: could not delete the timeline because it doesn't exist

HasChildren: could not delete the timeline because it has children

StopUploadQueue: ???

Other(#[from] anyhow::Error). could not delete timeline because something else, unexpected happened 

What would be the one-line description for StopUploadQueue? Something like "could not delete timeline because it is in Broken state" ? Can we make it more explicit and call the error `TimelineIsBroken`? Are there some other scenarios?

Does that mean that you cannot delete a broken timeline? We can live with that, but it would be nice if you could, so that you could fix the situation by deleting the broken timeline.

Reply:
The problem with the ambiguity is that, in timeline_delete, in the place where we call `remote_client.stop()`, we don't know why the queue is uninitialized. We just know that it _is_ uninitialized.

One reason why it can be uninitialized is that we failed to download the IndexParts during tenant load.
But, that is not exhaustive.
So, `TenantState::Broken` does not imply `upload queue is uninitialized`.

So, it seemed easier to keep things ambiguous in the error message and,
in the very unlikely event that it happens, check the logs to reconstruct why the queue was uninitialized.

We have no distinguished treatment of the `StopUploadQueue` case anyway, so, I can also shove it into `Other` if that's what you prefer.

Reply:
> The problem with the ambiguity is that, in timeline_delete, in the place where we call `remote_client.stop()`, we don't know why the queue is uninitialized. We just know that it _is_ uninitialized.

I see.

> One reason why it can be uninitialized is that we failed to download the IndexParts during tenant load. But, that is not exhaustive. So, `TenantState::Broken` does not imply `upload queue is uninitialized`.

I suppose we could check for that case, and say "tenant is broken" if it is, and "upload queue is uninitialized, which we didn't expect" otherwise. But if I understand correctly, we don't really expect that case to happen either? Can you call delete_timeline on a broken tenant?

> So, it seemed easier to keep things ambiguous in the error message and, in the very unlikely event that it happens, check the logs to reconstruct why the queue was uninitialized.
> 
> We have no distinguished treatment of the `StopUploadQueue` case anyway, so, I can also shove it into `Other` if that's what you prefer.

Ok. I'd go with Other, then, if they both just mean "something unexpected happened and we don't know how to proceed". With a different error code, someone who's working on the caller will waste time thinking if they need to do something differently with `StopUploadQueue` and `Other`.